### PR TITLE
Add benchmark perf harness for RyuJIT: RunBenchmarks

### PR DESCRIPTION
**Pull request by chuck-mitchell**
This is the initial pull request for the RyuJIT perf harness.  It supports benchmarking the current set of benchmarks under coreclr/tests/src/JIT/Performance/CodeQuality using the control file coreclr_benchmarks.xml.   There is also a legacy control file benchmarks.xml used by previous versions of RyuJIT running on Windows and Desktop CLR.

### CODE REVIEW

File: tests/src/RunBenchmarks/RunBenchmarks.cmd [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+set BENCHMARK_CONTROLS=-v -w -n 5
+set BENCHMARK_CONTROLS=-v -n 1 -s Bytemark
+set BENCHMARK_SET=-f %BENCHMARK_HOME%\coreclr_benchmarks.xml -notags broken
+set BENCHMARK_SWITCHES=%BENCHMARK_CONTROLS% -r %BENCHMARK_ROOT_DIR%
+
+%BENCHMARK_HOST% %BENCHMARK_RUNNER% %BENCHMARK_SET% %BENCHMARK_SWITCHES%
+
+taskkill /im DHandler.exe
+
+endlocal
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cmd]: In general, I think scripts should be as well documented as code. I'd like to see at least a header comment that describes the purpose and usage of the script.

This script obviously has hard-coded paths, so presumably won't work for most people. Can those paths be made relative, e.g., relative to the script itself (using %~dp0)?

BENCHMARK_CONTROLS is duplicated: delete the first?

Should you add @echo off? or do you want the commands echoed?

File: tests/src/RunBenchmarks/RunBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,1049 @@
...
+        public List<Benchmark>
+            SelectedBenchmarkList = new List<Benchmark>();
+        public List<Results>
+            ResultsList = new List<Results>();
+
+        public int NumberOfBenchmarksRun;
+        public int Failures = 0;
+
+        public Char[] ListSeparatorCharSet = new Char[] { ',', ' ' };
+        public char[] DirectorySeparatorCharSet = new char[] { '/', '\\' };
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: Mix of Char and char (casing). Probably should stick with lower-case.

File: tests/src/RunBenchmarks/RunBenchmarks.cs [hunk 3]
```diff
@@ -0,0 +1,1049 @@
...
+                {
+                    string arg = args[i++];
+                    string benchmark;
+                    string[] tags;
+                    string runner;
+
+                    switch (arg)
+                    {
+                        case "-n":
+                            arg = args[i++];
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: you'll obviously throw an exception if someone passes "-n" without a subsequent argument.

File: tests/src/RunBenchmarks/RunBenchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,1049 @@
...
+            BenchmarkList.Add(benchmark);
+
+            if (!BenchmarkSuiteTable.TryGetValue(suiteName, out benchmarkSuite))
+            {
+                benchmarkSuite = new BenchmarkSuite(suiteName);
+                BenchmarkSuiteTable.Add(suiteName, benchmarkSuite);
+            }
+            benchmarkSuite.BenchmarkList.Add(benchmark);
+
+            String[] tagList = tags.Split(ListSeparatorCharSet, StringSplitOptions.RemoveEmptyEntries);
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: seems like you should stick to lower-case 'string'

File: tests/src/RunBenchmarks/RunBenchmarks.cs [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+                }
+                if (useAVX)
+                {
+                    startInfo.Environment["COMPLUS_FeatureSIMD"] = "1";
+                    startInfo.Environment["COMPLUS_EnableAVX"] = "1";
+                }
+                startInfo.Environment["COMPLUS_gcConcurrent"] = "0";
+                startInfo.Environment["COMPLUS_gcServer"] = "0";
+                startInfo.Environment["COMPLUS_NoGuiOnAssert"] = "1";
+                startInfo.Environment["COMPLUS_BreakOnUncaughtException"] = "0";
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: For these to work on non-Windows as well (is that a goal?), you should use the "proper" casing of the environment variables: COMPlus_*

Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: That is to say, *nix is case-sensitive. that means using COMPlus_Version and COMPlus_DefaultVersion as well. Might want to double-check the others.

File: tests/src/RunBenchmarks/RunBenchmarks.cmd [hunk 6]
```diff
@@ -0,0 +1,20 @@
+setlocal
+
+set BENCHMARK_ROOT_DIR=\git\repos\DotNet\coreclr\bin\tests\Windows_NT.x64.Release\JIT\Performance\CodeQuality
+set BENCHMARK_HOME=\git\repos\DotNet\coreclr\tests\src\RunBenchmarks
+set BENCHMARK_HOST=%BENCHMARK_HOME%\bin\Debug\dnxcore50\RunBenchmarks.exe
+set BENCHMARK_HOST=%BENCHMARK_HOME%\bin\Debug\desktop\RunBenchmarks.exe
+set BENCHMARK_RUNNER=-runner CoreRun.exe
+
+start DHandler.exe
```
Comment (AndyAyersMS) [inline tests/src/RunBenchmarks/RunBenchmarks.cmd]: `DHandler` and (a bit lower down) `taskkill` are probably not available for most users.

They're probably also not really required, at least for CoreCLR, provided you set `COMPlus_noguionassert`.

File: tests/src/RunBenchmarks/benchmarks.xml [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+                <args>/nologo /optimize+ /noconfig @Repro.rsp</args>
+            </benchmark>
+    </benchmark-suite>
+    <benchmark-suite>
+        <name>System.XML</name>
+            <benchmark>
+                <name>System.XML</name>
+                <directory>.\System.XML\x86ret</directory>
+                <executable>SimpleTest.exe</executable>
+                <args>2000 &gt; out</args>
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/benchmarks.xml]: I wonder if it's a good idea to include stdout/stderr redirection in the 'args' option. I wonder if that's very portable. Perhaps, instead, there should be a <stdoutredirection> and maybe <stderrredirection> option, that, if set to true, causes the harness to automatically create the proper redirection.

File: tests/src/RunBenchmarks/benchmarks.xml [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+                <directory>.\mono</directory>
+                <executable>fastaredux.csharp.exe</executable>
+                <args>25000000 &gt; out</args>
+                <run-in-shell>true</run-in-shell>
+            </benchmark>
+            <benchmark>
+                <name>mono-knucleotide</name>
+                <directory>.\mono</directory>
+                <executable>knucleotide.csharp-4.exe</executable>
+                <args>0 &lt; knucleotide-input25000000.txt &gt; out</args>
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/benchmarks.xml]: So stdin redirection is the only way to pass the input file? I wonder if my stdout comments apply here, too.

File: tests/src/RunBenchmarks/RunBenchmarks.cs [hunk 9]
```diff
@@ -0,0 +1,1049 @@
...
+// <benchmark-system>
+//     <benchmark-root-directory>ROOT_DIRECTORY</benchmark-root-directory> // optional, can be on command line
+//     <benchmark-suite>
+//         <name>SUITE_NAME</name>
+//             <benchmark>
+//                 <name>BENCHMARK_NAME</name>
+//                 <directory>BENCHMARK_DIRECTORY</directory>
+//                 <executable>EXECUTABLE_PATH</executable>
+//                 <args>EXECUTABLE_ARGUMENTS</args>           // optional args, can redirect output: &gt; > out
+//                 <run-in-shell>true</run-in-shell>           // optional
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cs]: You should describe this.

File: tests/src/RunBenchmarks/build.cmd [hunk 10]
```diff
@@ -0,0 +1,10 @@
+rem *** Build for Desktop CLR
+
+mkdir bin\Debug\desktop
+csc /define:DESKTOP /nologo /debug /target:exe /out:bin\Debug\desktop\RunBenchmarks.exe RunBenchmarks.cs 
+
+rem *** Build for Core CLR
+
+dotnet restore
+dotnet compile
+
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/build.cmd]: Same comment as before: scripts should be as well-commented as source code.

Should you add setlocal?

Should the Desktop/CoreCLR builds be in different scripts?
Should you check for and report errors?
Should you report where the results were generated?

File: tests/src/RunBenchmarks/build.sh [hunk 11]
```diff
@@ -0,0 +1,2 @@
+dotnet restore
+dotnet compile
```
Comment (BruceForstall) [inline tests/src/RunBenchmarks/build.sh]: Same comments as for build.cmd

**Comment by BruceForstall** [comment]
In general, it looks great. It's nice that it's in C# instead of the JScript we've been using!

File: tests/src/RunBenchmarks/RunBenchmarks.cmd (hunk 6 above)
Comment (BruceForstall) [inline tests/src/RunBenchmarks/RunBenchmarks.cmd]: taskkill comes with Windows. Maybe we should look into open sourcing DHandler? Agree about setting COMPlus_NoGuiOnAssert

Comment (AndyAyersMS) [inline tests/src/RunBenchmarks/RunBenchmarks.cmd]: `DHandler` is super-annoying, I'd say we should not bother. There's less likelihood of GUI popups on coreclr anyways.

**Pull request closed without merging**